### ISSUE TITLE
pileup on streaming SAM records

### ISSUE THREAD

**Issue by winni2k** [pysam-developers/pysam#915]
I have a use case where I want to implement a read-group-aware `samtools mpileup` on a file streamed from stdin or a UNIX file descriptor. I think this could be done with pysam's pileup engine, but my testing shows that pileup requires an indexed BAM. Is there an easy way to get the pileup to work on SAM/BAM records streamed from stdin? If we assume that the reads are position sorted, then this should be feasible, no?

**Comment by jmarshall** [comment]
The `samtools mpileup` command works fine when reading SAM from stdin, so certainly the underlying HTSlib pileup engine works with such input.

What pysam code have you considered, and did it result in a “_pileup of samfiles not implemented yet_” exception?

**Comment by winni2k** [comment]
Here is what I ran. `test2.bam` has no index. I am not asking for a particular region in `pileup`, so an index should not be necessary. Likewise, if I am pileing up from stdin, then an index won't be available either.

Is there another way to do this?

```
In [1]: import pysam

In [2]: f = pysam.AlignmentFile('test2.bam', 'rb')

In [3]: for r in f.pileup():
   ...:     r
   ...:
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-387591687760> in <module>
----> 1 for r in f.pileup():
      2     r
      3

pysam/libcalignmentfile.pyx in pysam.libcalignmentfile.AlignmentFile.pileup()

ValueError: no index available for pileup
```

This is running pysam 0.15.3

**Comment by jmarshall** [comment]
Try applying PR #916 and rebuilding pysam.

**Comment by winni2k** [comment]
Nice! I'll have a go.

**Comment by PanZiwei** [comment]
Hi @jmarshall ,
I still got the error `ValueError: no index available for pileup` with pysam==0.16.0.1 when trying to use another tool which call pysam inside. Here is the error message:
```
File "pysam/libcalignmentfile.pyx", line 1331, in pysam.libcalignmentfile.AlignmentFile.pileup
ValueError: no index available for pileup
```
 do you have an idea on this issue?

Thank you so much for your help!
